tests/phpunit/elementor/modules/atomic-widgets/test-styles-renderer.php

```diff
@@ -0,0 +1,334 @@
+        // Arrange.
+        $styles = [
+            [
+                'id' => 'test-style',
+                'type' => 'class',
+                'variants' => [
+                    [
+                        'props' => [
+                            'color' => 'green',
+                        ],
+                        'meta' => [
+                            'breakpoint' => 'mobile',
+                        ],
+                    ],
+					[
+						'props' => [
+							'color' => 'blue',
+						],
+						'meta' => [
+							'breakpoint' => 'tablet', // non-existing breakpoint
+						],
+					]
+                ],
+            ],
+        ];
+        $transformers = [];
+        $breakpoints = [
+            'mobile' => [
+                'type' => 'max-width',
+                'width' => 768,
+            ],
+        ];
+
+        $styleRender = new Styles_Renderer( $transformers, $breakpoints );
+
+        // Act.
+        $css = $styleRender->render( $styles );
+
+        // Assert.
+		$this->assertMatchesSnapshot( $css );
+    }
+
+    public function test_styles_renderer__style_with_transformers() {
+        // Arrange.
+        $styles = [
+            [
+                'id' => 'test-style',
+                'type' => 'class',
+                'variants' => [
+                    [
+                        'props' => [
+                            'font-size' => [
+								'$$type' => 'size',
+								'value' => [
+									'size' => 14,
+									'unit' => 'px'
+								]
+							],
+							'text-decoration' => [
+								'$$type' => 'text-decoration', // non-existing transformer
+								'value' => 'underline'
+							],
+							'box-shadow' => [
+								'$$type' => 'array',
+								'value' => [
+									'x' => 0,
+									'y' => 0,
+									'blur' => 5,
+									'color' => [
+										'$$type' => 'color', // nested transformable value - evaluated within the array transformer function
+										'value' => '000000'
+									],
+								]
+							]
+                        ],
+                        'meta' => [],
+                    ],
+                ],
+            ],
+        ];
+        $transformers = [
+            'size' => function( $value ) {
+				$size = (int)$value['size'];
+				$unit = $value['unit'];
+				return $size . $unit;
+            },
+			'color' => function( $value ) {
+				return '#' . $value;
+			},
+			'array' => function( $value, $transform ) {
+				$x = $value['x'];
+				$y = $value['y'];
+				$blur = $value['blur'];
+				$color = $transform( $value['color'] );
+
+				return $x.' '.$y.' '.$blur.' '.$transform( $color );
+			}
+        ];
+        $breakpoints = [];
+
+        $styleRender = new Styles_Renderer( $transformers, $breakpoints );
+
+        // Act.
+        $css = $styleRender->render( $styles );
+
+        // Assert.
+		$this->assertMatchesSnapshot( $css );
+    }
+
+	public function test_styles_renderer__multiple_style_definitions_and_variants() {
```

Comment:
isn't it re-testing a lot of the already tested cases?

Reply:
Cleaned it up a bit to only focus on the multiple styles and variants